src/client/Client.cc

```diff
@@ -9820,21 +9820,297 @@ int Client::preadv(int fd, const struct iovec *iov, int iovcnt, loff_t offset)
 {
   if (iovcnt < 0)
     return -CEPHFS_EINVAL;
-  return _preadv_pwritev(fd, iov, iovcnt, offset, false);
+  return _preadv_pwritev(fd, iov, iovcnt, offset, false, nullptr);
 }
 
-int64_t Client::_read(Fh *f, int64_t offset, uint64_t size, bufferlist *bl)
+class Client_Read_Finisher : public Context {
+public:
+  bool iofinished;
+  void finish_io(int r);
+  void finish_onuninline(int r);
+  void finish_ui(int r);
+  Client_Read_Finisher(Client *clnt, Context *onfinish, Context *onuninline,
+                       Context *iofinish, bool is_read_async, int have,
+                       bool movepos, utime_t start, Fh *f, Inode *in,
+                       uint64_t fpos, int64_t offset, uint64_t size)
+    : clnt(clnt), onfinish(onfinish), onuninline(onuninline),
+      iofinish(iofinish), is_read_async(is_read_async), have(have),
+      f(f), in(in), start(start), fpos(fpos), offset(offset),
+      size(size), movepos(movepos) {
+    iofinished_r = 0;
+    onuninlinefinished_r = 0;
+    iofinished = false;
+    onuninlinefinished = onuninline == nullptr;
+  }
+
+  void finish(int r) override {
+    //
+    delete this;
+  }
+
+private:
+  Client *clnt;
+  Context *onfinish;
+  Context *onuninline;
+  Context *iofinish;
+  bool is_read_async;
+  int have;
+  Fh *f;
+  Inode *in;
+  utime_t start;
+  uint64_t fpos;
+  int64_t offset;
+  uint64_t size;
+  int64_t iofinished_r;
+  int64_t onuninlinefinished_r;
+  bool onuninlinefinished;
+  bool movepos;
+  bool try_complete();
+};
```

Comment:
We'd better put this in the `Client.h`, the `Client.cc` is already very huge one ?

And also for `Client_Read_Sync_Async` ?